lib/io_intf.ml

```diff
@@ -0,0 +1,84 @@
+module Tls_error = struct
+  type t =
+    | Tls_alert of Packet.alert_type
+    (** [Tls_alert] exception received from the other endpoint *)
+    | Tls_failure of Engine.failure
+    (** [Tls_failure] exception while processing incoming data *)
+    | Connection_closed
+    | Connection_not_ready
+    | Unexpected_eof
+    | Unable_to_renegotiate
+    | Unable_to_update_key
+  [@@deriving sexp_of]
+end
+
+module type Concurrency = sig
+  type 'a t
+
+  val bind : 'a t -> f:('a -> 'b t) -> 'b t
+  val return : 'a -> 'a t
+  val map : 'a t -> f:('a -> 'b) -> 'b t
+
+  val tls_error : Tls_error.t -> 'a t
+end
+
+module type Fd = sig
+  module Concurrency : Concurrency
```

Comment:
Isn't there already a module type in Core for bind/return/map that we could use, instead of defining a new Concurrency module signature here?  It's not quite clear to me what `tls_error` is doing in the Concurrency module type either.

Reply:
I don't think lib/ has a dependency on Core, right? I didn't want to impose opinions about which standard library to use.

Reply:
Right -- I think you're better off moving all the `lib/` stuff directly into `async/`. Experience has shown that functorising over Lwt/Async doesn't work well, and there are quite deep differences in how both handle concurrency (e.g. Monitors _vs_ monadic exceptions). 

Depending on your plans for the async backend, it might be better for you to keep this outside of the mirage/ org, so it can be released independently with the rest of the Core releases.  Have you considered unifying the async_ssl and this interface into one library?  If it's better to keep it in this repository, we can, but the tls library is structured such that you should be able to use the core of it externally too.

Reply:
I've moved this code out of lib/ and into async/